### PR TITLE
implement lenient parser

### LINKED ISSUE #5: Add a lenient mode to QueryParser
The role of the `QueryParser` is to give an off-the-shelf solution to people wanting to deliver a search box to end user.
Currently `QueryParser::parse` can return an error if the query does not match our query grammar.

e.g.: `+(happy`  : the parenthesis is not closed.

This is great, but how should somebody writing a site handle this error?
Maybe for a log analysis search engine, displaying an explicit _syntax error_ is a good idea, but for most use case, (ecommerce, document search, etc.) we want to never return an error and just do our best to handle the user query.

For this reason, we want to add a new `parse_lenient` that would never fail and always return a result.

```rust
pub fn parse_query_lenient(&self, query: &str) -> Box<Query> {
    // ...
}
``` 

The behavior of this lenient mode does NOT have to be very smart...
For instance, it could be a second very naive parser taking over when the initial parser failed.

For instance, this naive parser could remove all special characters and run the initial parser. 

Smarter attempts to interpret the faulty user query are also welcome, as long as we return `Box<Query>` and we are confident the code will never panic.

### PR DESCRIPTION
**Pull request by trinity-1686a**
fix #5 

Most of the grammar part is done, however it's not reporting hints just yet (this isn't really required, but would be a nice addition, to help someone understand how their query was understood, and why the parser think it's somewhat wrong).

At the moment, only the part in query-grammar is done. More work is required to convert the AST to a query without failing. This part hasn't been explored yet, but should be before the ticket is considered fixed.

I'm sending this as a draft to get initial comments on the parser part

### CODE REVIEW

**Comment by adamreichold** [comment]
The main question that came to mind reading this was whether the lenient parser should fully replace the existing parser. I think as an API between systems, the existing - let's say "strict" - parser is still useful. Of course, as a user interface, a lenient parser is very much preferable.

How do people think about having both parsers and choosing between them at compile time or via a runtime flag? I understand that maintaining two parsers is a significant commitment but I also feel that replacing the existing behaviour in a "flag day" type of change could trigger a certain amount of user backlash.

**Comment by trinity-1686a** [comment]
the goal isn't to remove the strict parser. It was rewritten so both could share some code, but it will definitely be kept for the foreseeable future.

**Comment by adamreichold** [comment]
> the goal isn't to remove the strict parser. It was rewritten so both could share some code, but it will definitely be kept for the foreseeable future.

Ah sorry, then I misunderstood the PR in my first reading and the concern does not apply at all. Thanks for clarifying!

File: query-grammar/src/query_grammar.rs [hunk 1]
```diff
@@ -20,249 +19,359 @@ use crate::Occur;
...
-/// need to be escaped with a backslash character '\'.
-fn field_name<'a>() -> impl Parser<&'a str, Output = String> {
-    static ESCAPED_SPECIAL_CHARS_RE: Lazy<Regex> =
-        Lazy::new(|| Regex::new(ESCAPED_SPECIAL_CHARS_PATTERN).unwrap());
-
-    recognize::<String, _, _>(escaped(
-        (
-            take_while1(|c| !SPECIAL_CHARS.contains(&c) && c != '-'),
-            take_while(|c| !SPECIAL_CHARS.contains(&c)),
+/// consume a field name followed by collon. Return the field name with escape sequence
```
Comment (fulmicoton) [inline query-grammar/src/query_grammar.rs]:
```suggestion
/// consume a field name followed by colon. Return the field name with escape sequence
```

File: query-grammar/src/query_grammar.rs [hunk 2]
```diff
@@ -488,56 +888,67 @@ mod test {
...
+        let query_lenient = parse_to_ast_lenient(query).0;
+        let query_lenient_str = format!("{query_lenient:?}");
+        assert_eq!(query_lenient_str, expected, "lenient parser failed");
     }
 
-    fn test_is_parse_err(query: &str) {
-        assert!(parse_to_ast().parse(query).is_err());
+    #[track_caller]
+    fn test_is_parse_err(query: &str, lenient_expected: &str) {
+        assert!(parse_to_ast(query).is_err(), "strict parser succeeded");
```
Comment (fulmicoton) [inline query-grammar/src/query_grammar.rs]:
```suggestion
        assert!(parse_to_ast(query).is_err(), "strict parser succeeded where an error was expected.");
```

File: query-grammar/src/lib.rs [hunk 3]
```diff
@@ -1,19 +1,25 @@
...
 
+/// Parse a query
 pub fn parse_query(query: &str) -> Result<UserInputAst, Error> {
-    let (user_input_ast, _remaining) = parse_to_ast().parse(query).map_err(|_| Error)?;
+    let (_remaining, user_input_ast) = parse_to_ast(query).map_err(|_| Error)?;
     Ok(user_input_ast)
 }
+
+/// Parse a query, trying to recover from syntax errors, and giving hints toward fixing errors.
+pub fn parse_query_leniet(query: &str) -> (UserInputAst, Vec<(usize, String)>) {
```
Comment (fulmicoton) [inline query-grammar/src/lib.rs]:
```suggestion
pub fn parse_query_lenient(query: &str) -> (UserInputAst, Vec<(usize, String)>) {
```

File: query-grammar/src/query_grammar.rs [hunk 4]
```diff
@@ -413,38 +735,121 @@ fn aggregate_binary_expressions(
...
+    map(delimited(space0, opt(ast), eof), |opt_ast| {
+        rewrite_ast(opt_ast.unwrap_or_else(UserInputAst::empty_query))
+    })(i)
+}
+
+pub fn parse_to_ast_lenient(i: &str) -> (UserInputAst, ErrorList) {
+    // TODO remove front space and detect final eof.
+    let (res, mut error) = ast_infallible(i).unwrap().1;
+
+    // convert end-based index to start-based index.
```
Comment (fulmicoton) [inline query-grammar/src/query_grammar.rs:956]: for loop please.

```rust
for error in errors {
     error.0 = query_str.len() - error.0;
}
```

File: query-grammar/src/query_grammar.rs [hunk 5]
```diff
@@ -413,38 +735,121 @@ fn aggregate_binary_expressions(
...
-        .map(rewrite_ast)
+pub fn parse_to_ast(i: &str) -> IResult<&str, UserInputAst> {
+    map(delimited(space0, opt(ast), eof), |opt_ast| {
+        rewrite_ast(opt_ast.unwrap_or_else(UserInputAst::empty_query))
+    })(i)
+}
+
+pub fn parse_to_ast_lenient(i: &str) -> (UserInputAst, ErrorList) {
+    // TODO remove front space and detect final eof.
+    let (res, mut error) = ast_infallible(i).unwrap().1;
```
Comment (fulmicoton) [inline query-grammar/src/query_grammar.rs]:
```suggestion
    let (res, mut errors) = ast_infallible(i).unwrap().1;
```

File: query-grammar/src/query_grammar.rs [hunk 6]
```diff
@@ -413,38 +735,121 @@ fn aggregate_binary_expressions(
...
-        .with(optional(ast()).skip(eof()))
-        .map(|opt_ast| opt_ast.unwrap_or_else(UserInputAst::empty_query))
-        .map(rewrite_ast)
+pub fn parse_to_ast(i: &str) -> IResult<&str, UserInputAst> {
+    map(delimited(space0, opt(ast), eof), |opt_ast| {
+        rewrite_ast(opt_ast.unwrap_or_else(UserInputAst::empty_query))
+    })(i)
+}
+
+pub fn parse_to_ast_lenient(i: &str) -> (UserInputAst, ErrorList) {
```
Comment (fulmicoton) [inline query-grammar/src/query_grammar.rs]:
```suggestion
pub fn parse_to_ast_lenient(query_str: &str) -> (UserInputAst, ErrorList) {
```

File: query-grammar/src/infallible.rs [hunk 7]
```diff
@@ -0,0 +1,286 @@
+//! nom combinators for infallible operations
+
+use nom::{IResult, InputLength};
+pub type ErrorList = Vec<(usize, String)>;
```
Comment (fulmicoton) [inline query-grammar/src/infallible.rs]: No clue what the first `usize` argument is, nor the String.

Comment (fulmicoton) [reply query-grammar/src/infallible.rs]: Can we introduce type to make it less confusing?

It seems like the first argument changes meaning halfway through. (offset to the end to offset from the start)

File: query-grammar/src/infallible.rs [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+                res.push(o);
+                i = i1;
+            }
+        }
+
+        loop {
+            let len = i.input_len();
+            match sep.parse(i.clone()) {
+                Err(_) => unreachable!(),
+                Ok((i1, (_, mut err))) => {
```
Comment (fulmicoton) [inline query-grammar/src/infallible.rs]: The `let .. else ...` pattern is a good candidate here.
```
let Ok((i1, (_, mut err))) = sep.parse(i.clone()) else { unreachable!() };
```

You can also add a function

```
fn unwrap_unfallible<T>(res: Result<T, Infallible>) -> T {}
```

if it ends up useful in several places.

File: query-grammar/src/infallible.rs [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+    mut f: F,
+) -> impl FnMut(I) -> JResult<I, Vec<O>>
+where
+    I: Clone + InputLength,
+    F: nom::Parser<I, (O, ErrorList), std::convert::Infallible>,
+    G: nom::Parser<I, (O2, ErrorList), std::convert::Infallible>,
+{
+    move |mut i: I| {
+        let mut res = Vec::new();
+        let mut errors = Vec::new();
```
Comment (fulmicoton) [inline query-grammar/src/infallible.rs]: can you add the type of these two please?

File: query-grammar/src/infallible.rs [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+
+        loop {
+            let len = i.input_len();
+            match sep.parse(i.clone()) {
+                Err(_) => unreachable!(),
+                Ok((i1, (_, mut err))) => {
+                    errors.append(&mut err);
+
+                    match f.parse(i1.clone()) {
+                        Err(_) => unreachable!(),
```
Comment (fulmicoton) [inline query-grammar/src/infallible.rs]: same as above.

File: query-grammar/src/query_grammar.rs [hunk 11]
```diff
@@ -413,38 +735,121 @@ fn aggregate_binary_expressions(
...
+            .map(|(_, occur, ast)| (occur, ast))
+            .collect();
+        return (UserInputAst::Clause(clauses), Vec::new());
+    }
+
+    if leafs.iter().all(|(_, occur, _)| occur.is_none()) {
+        let first = leafs.remove(0).2;
+        let rest = leafs
+            .into_iter()
+            .map(|(operand, _, ast)| (operand.unwrap_or(BinaryOperand::Or), ast))
```
Comment (fulmicoton) [inline query-grammar/src/query_grammar.rs]: should the operand depend on the query parser default behavior?

**Review by fulmicoton — CHANGES_REQUESTED** [review]
Nice approach. 

But you really should put more effort in making your code readable.
Here are a bunch of concrete things you can do to improve this.


# Have proper variable names

This is short and well contained, so `i` and `o` might be ok variable names.
I would have preferred `input` and `output` but I guess this can be argued.
`i1`, `i2` however, is really not helpful.

# Stop abusing iterator style. 

For instance, if you do it to mutate stuff, that's probably a bad idea.
If you start using rare iterator functions, that's probably a bad idea too.
If you end up struggling with dealing with error handling, using more than one trick like transpose and `collect::<Result<T,_>>`, it is probably a bad idea.
If your code seems to suggest a variable is a vec but it is in fact an Option or a Result, this is a bad idea.

# Consider using struct instead of tuple

`type Error = (usize, String);`
That kind of tuple makes code easy to write, but hard to understand.
You actually changed the semantics of the first argument, which is making things even worse.

If possible, I would have preferred two structs.
This however comes with a verbosity cost, so this may not be pertinent:  I leave you judge of whether the change should be done or not.

# Comment more

Something like
`type Error = (usize, String);`
is really hard for the reader.

# Show types in your code when it is not easy to guess.

We sometimes have to read rust without type hints.
For instance in github or in an editor for which rust analyzer has not finished crunching a
freshly checked out branch.
It is useful to add explicit types here in strategic places.

File: query-grammar/src/query_grammar.rs (hunk 11 above)
Comment (trinity-1686a) [reply query-grammar/src/query_grammar.rs]: this part needs to be improved to work with a mix of binary operand and occurs, however I'm yet to find how to mix both in a way that makes sens and transition coherently from full-occurs and full-binary-operand

**Review by fulmicoton — APPROVED** [review]

**Pull request merged**